Objects/dictobject.c

```diff
@@ -2591,6 +2581,102 @@ dict_merge(PyObject *a, PyObject *b, int override)
     return 0;
 }
 
+/* Implementation of d.update(obj). There are two cases:
+ * - if obj has a "keys" method, assume that it is a mapping
+ * - otherwise, obj must be an iterable of (key, value) pairs
+ */
+static int
+dict_merge_from_any(PyObject *self, PyObject *obj)
+{
+    /* Fast path for dict subclasses */
+    if (PyDict_Check(obj)) {
+        return dict_merge(self, obj, 1);
+    }
+
+    _Py_IDENTIFIER(keys);
+    PyObject *func;
+    if (_PyObject_LookupAttrId(obj, &PyId_keys, &func) < 0) {
```

Comment:
That sounds redundant with dict_merge() which calls PyMapping_Keys(). Maybe dict_merge() could be modified to dispatch to PyDict_MergeFromSeq2() if PyMapping_Keys() keys with an attribute error, rather than trying to get the attribute twice?

It's not directly related to your change, it can be done/discussed in a separated PR.

Reply:
I agree, I also noticed this. There is some room for improvement here. Let's do it in a new PR after this is merged.